analyzers/src/SonarAnalyzer.CFG/Helpers/OperationWalker.cs

```diff
@@ -0,0 +1,66 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2021 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Collections.Generic;
+using Microsoft.CodeAnalysis;
+using StyleCop.Analyzers.Lightup;
+
+namespace SonarAnalyzer.CFG.Helpers
+{
+    public abstract class OperationWalker : OperationWalker<object>
+    {
+        protected abstract void VoidVisitOperation(IOperation operation);
+
+        protected override object VisitOperation(IOperation operation)
+        {
+            VoidVisitOperation(operation);
+            return null;
+        }
+    }
+
+    public abstract class OperationWalker<TResult> : OperationVisitor<TResult>
```

Comment:
I tried to improve the code and then I realized that this is actually `OperationFinder`, because it finds first and returns that. So let's write these in a different way:

First general walker:
```
    public abstract class OperationWalker
    {
        protected abstract void DefaultVisit(IOperationWrapperSonar operation);

        protected virtual bool VisitChildren(IOperationWrapperSonar operation) => true;

        public void Visit(IOperationWrapperSonar operation) =>
            Visit(operation.Instance);

        public void Visit(IOperation operation)
        {
            if (operation == null)
            {
                return;
            }
            var queue = new Queue<IOperation>();
            queue.Enqueue(operation);
            while (queue.Any())
            {
                var current = new IOperationWrapperSonar(queue.Dequeue());
                DefaultVisit(current);
                if (VisitChildren(current))
                {
                    foreach (var child in current.Children)
                    {
                        queue.Enqueue(child);
                    }
                }
            }
        }
    }
```
Then finder on top of it:
```
    public abstract class OperationFinder<TResult> : OperationWalker
        where TResult : class
    {
        private TResult result;

        protected abstract TResult ReturnValue(IOperationWrapperSonar operation);

        public TResult FindFirst(IOperation operation)
        {
            result = null;
            Visit(operation);
            return result;
        }

        protected override bool VisitChildren(IOperationWrapperSonar operation) =>
            result == null;

        protected override void DefaultVisit(IOperationWrapperSonar operation) =>
            result ??= ReturnValue(operation);
    }
```
This is not thread-safe, but we can do this if we need to:
```
    public abstract class OperationFinder<TResult> : OperationWalker
        where TResult : class
    {
        private readonly IOperationWrapperSonar root;
        private TResult result;

        protected abstract TResult ReturnValue(IOperationWrapperSonar operation);

        protected OperationFinder(IOperationWrapperSonar root) =>
            this.root = root;

        public TResult FindFirst()
        {
            Visit(root);
            return result;
        }

        protected override bool VisitChildren(IOperationWrapperSonar operation) =>
            result == null;

        protected override void DefaultVisit(IOperationWrapperSonar operation) =>
            result ??= ReturnValue(operation);
    }
```

Reply:
OperationWalker you've suggested does not stop on the first found operation, instead it will continue to process what's left in the queue. After a long brainstorming session with @csaba-sagi-sonarsource we've found another interesting approach.

Reply:
But thanks about wrapper inspirations!